File: modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/UnaryExpression.java
/*
*  Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
*
*  WSO2 Inc. licenses this file to you under the Apache License,
*  Version 2.0 (the "License"); you may not use this file except
*  in compliance with the License.
*  You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing,
*  software distributed under the License is distributed on an
*  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
*  KIND, either express or implied.  See the License for the
*  specific language governing permissions and limitations
*  under the License.
*/
package org.ballerinalang.model.expressions;

import org.ballerinalang.model.NodeExecutor;
import org.ballerinalang.model.NodeLocation;
import org.ballerinalang.model.NodeVisitor;
import org.ballerinalang.model.Operator;
import org.ballerinalang.model.WhiteSpaceDescriptor;
import org.ballerinalang.model.values.BBoolean;
import org.ballerinalang.model.values.BFloat;
import org.ballerinalang.model.values.BInteger;
import org.ballerinalang.model.values.BValue;
import org.ballerinalang.model.values.BValueType;

import java.util.function.BiFunction;

/**
 * {@code UnaryExpression} represents a unary expression.
 *
 * @since 0.8.0
 */
public class UnaryExpression extends AbstractExpression {

    protected Operator op;
    protected Expression rExpr;
    //ToDO this has to be improved property since Unary does not need BiFunction
    private BiFunction<BValue, BValue, BValue> evalFuncNewNew;

    public static final BiFunction<BValue, BValue, BValue> NOT_BOOLEAN_FUNC =
            (lVal, rVal) -> new BBoolean(!((BValueType) rVal).booleanValue());

    public static final BiFunction<BValue, BValue, BValue> NEGATIVE_INT_FUNC =
            (lVal, rVal) -> new BInteger(-(((BValueType) rVal).intValue()));

    public static final BiFunction<BValue, BValue, BValue> POSITIVE_INT_FUNC =
            (lVal, rVal) -> rVal;

    public static final BiFunction<BValue, BValue, BValue> NEGATIVE_FLOAT_FUNC =
            (lVal, rVal) -> new BFloat(-(((BValueType) rVal).floatValue()));

    public static final BiFunction<BValue, BValue, BValue> POSITIVE_FLOAT_FUNC =
            (lVal, rVal) -> rVal;

            
    public BiFunction<BValue, BValue, BValue> getEvalFunc() {
        return evalFuncNewNew;
    }

    public void setEvalFunc(BiFunction<BValue, BValue, BValue> evalFuncNewNew) {
        this.evalFuncNewNew = evalFuncNewNew;
    }
    
    public UnaryExpression(NodeLocation location, WhiteSpaceDescriptor whiteSpaceDescriptor, Operator op,
                           Expression rExpr) {
        super(location, whiteSpaceDescriptor);
        this.op = op;
        this.rExpr = rExpr;
    }

    public Expression getRExpr() {
        return rExpr;
    }

    public Operator getOperator() {
        return op;
    }

    @Override
    public void accept(NodeVisitor visitor) {
        visitor.visit(this);
    }

    @Override
    public BValue execute(NodeExecutor executor) {
        return executor.visit(this);
    }
}

File: modules/web/js/ballerina/ast/expressions/unary-expression.js
/**
 * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import _ from 'lodash';
import Expression from './expression';
import FragmentUtils from '../../utils/fragment-utils';
import ASTFactory from '../ast-factory.js';

/**
 * Construct for UnaryExpression
 * @param {Object} args - Arguments to create the Unary Expression
 * @constructor
 * @augments Expression
 * */
class UnaryExpression extends Expression {
    constructor(args) {
        super('UnaryExpression');
        this.whiteSpace.defaultDescriptor.regions = {
            0: '',
            1: '',
            2: ' ',
        };
        this._operator = _.get(args, 'operator');
    }

    /**
     * setting parameters from json
     * @param {Object} jsonNode to initialize from
     * */
    initFromJson(jsonNode) {
        if (!_.isNil(jsonNode.children[0])) {
            const rightExpression = ASTFactory.createFromJson(jsonNode.children[0]);
            rightExpression.setParent(this, { doSilently: true });
            rightExpression.initFromJson(jsonNode.children[0]);
            this.setRightExpression(rightExpression, { doSilently: true });
        }
    }

    /**
     * Set the expression from the expression string
     * @param {string} expressionString
     * @override
     */
    setExpressionFromString(expression, callback) {
        if (!_.isNil(expression)) {
            const fragment = FragmentUtils.createExpressionFragment(expression);
            const parsedJson = FragmentUtils.parseFragment(fragment);
            if ((!_.has(parsedJson, 'error')
                   || !_.has(parsedJson, 'syntax_errors'))) {
                this.initFromJson(parsedJson);
                if (_.isFunction(callback)) {
                    callback({ isValid: true });
                }
            } else if (_.isFunction(callback)) {
                callback({ isValid: false, response: parsedJson });
            }
        }
    }

    getExpressionString() {
        let expString = this.getOperator() + this.getWSRegion(1);
        expString += (!_.isNil(this.getRightExpression())) ? this.getRightExpression().getExpressionString() : '';
        return expString;
    }

    setOperator(operator, options) {
        this.setAttribute('_operator', operator, options);
    }

    getOperator() {
        return this._operator;
    }

    setRightExpression(rightExpression, options) {
        this.setAttribute('_rightExpression', rightExpression, options);
    }

    getRightExpression() {
        return this._rightExpression;
    }
}

export default UnaryExpression;
